Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp

```diff
@@ -1335,7 +1335,38 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
     Vector<GridTrack>& allTracks = tracks(m_direction);
 
     while (auto* gridItem = iterator.nextGridItem()) {
+
         bool isNewEntry = itemsSet.add(gridItem).isNewEntry;
+
+        // Masonry Track Sizing
+        // https://drafts.csswg.org/css-grid-3/#track-sizing
+        // We should only compute track sizes on a subset of items.
+        //
+        // - Items placed at the first implicit line in the masonry axis.
+        // - Items that have a specified definite placement in the grid axis.
+        // - Items that span all grid axis tracks.
+        if (m_renderGrid->areMasonryRows() || m_renderGrid->areMasonryColumns()) {
+            bool skipTrackSizing = true;
+            auto gridAxisDirection = m_renderGrid->areMasonryRows() ? ForColumns : ForRows;
+            auto masonryAxisDirection = m_renderGrid->areMasonryRows() ? ForRows : ForColumns;
+            auto span = GridPositionsResolver::resolveGridPositionsFromStyle(*m_renderGrid, *gridItem, gridAxisDirection);
+
+            // Items placed at the first implicit line in the masonry axis.
+            if (!m_renderGrid->gridSpanForChild(*gridItem, masonryAxisDirection).startLine())
+                skipTrackSizing = false;
+
+            // Items that have a specified definite placement in the grid axis.
+            if (!span.isIndefinite())
+                skipTrackSizing = false;
+
+            // Items that span all grid axis tracks.
+            if (!m_renderGrid->gridSpanForChild(*gridItem, masonryAxisDirection).startLine() && m_renderGrid->gridSpanForChild(*gridItem, gridAxisDirection).endLine() == tracks(m_direction).size())
```

Comment:
This should be gridAxisDirection for both calls, right?

Could also be `m_renderGrid->gridSpanForChild(*gridItem, gridAxisDirection).integerSpan() == tracks(m_direction).size()`